### PR TITLE
Make chunk encoding an explicit operation

### PR DESCRIPTION
**Pull request by bboreham**
Remove all the scary comments and make it more obvious where encoding happens

This should make the code in #1164 work again (has since been reverted)

### CODE REVIEW

**Comment by tomwilkie** [comment]
Agree the current state is a mess but its hard to ensure this will cover all the cases.  Could we make `Encoded` call `Encode` if its not already encoded?  Similarly have `Checksum` call `Encode` if its not...

**Comment by bboreham** [comment]
Just to check I understand: your preference is for a subtle performance degradation over a hard failure?
I guess I could live with that.

**Comment by bboreham** [comment]
There is no `Checksum()` afaics.  `ExternalKey()` seems like the main place where this would go wrong so I added a last-ditch `Encode()` there.

**Comment by bboreham** [comment]
Nope, that doesn't work: `Chunk` is used to supply key details, e.g. in `store.Get()`, so we can't call `Encode()`.

File: pkg/chunk/chunk_store_test.go [hunk 1]
```diff
@@ -430,8 +430,9 @@ func TestChunkStoreRandom(t *testing.T) {
 					ts,
 					ts.Add(chunkLen*time.Second),
 				)
-
-				err := store.Put(ctx, []Chunk{chunk})
+				err := chunk.Encode()
```
Comment (gouthamve) [inline pkg/chunk/chunk_store_test.go]: Why are we Encoding here before putting? afaics, it is not required?

File: pkg/chunk/chunk_test.go [hunk 2]
```diff
@@ -234,7 +236,9 @@ func BenchmarkDecode10000(b *testing.B) { benchmarkDecode(b, 10000) }
 
 func benchmarkDecode(b *testing.B, batchSize int) {
 	chunk := benchmarkChunk(model.Now())
-	buf, err := chunk.Encode()
+	err := chunk.Encode()
```
Comment (gouthamve) [inline pkg/chunk/chunk_test.go]: Same, we don't need to call `Encode()`?

File: pkg/chunk/chunk_test.go [hunk 3]
```diff
@@ -98,7 +98,9 @@ func TestChunkCodec(t *testing.T) {
 		},
 	} {
 		t.Run(fmt.Sprintf("[%d]", i), func(t *testing.T) {
-			encoded, err := c.chunk.Encode()
+			err := c.chunk.Encode()
```
Comment (gouthamve) [inline pkg/chunk/chunk_test.go]: Same here, no need to call `Encode()` here.

File: pkg/ingester/flush.go [hunk 4]
```diff
@@ -312,7 +312,11 @@ func (i *Ingester) flushChunks(ctx context.Context, fp model.Fingerprint, metric
 
 	wireChunks := make([]chunk.Chunk, 0, len(chunkDescs))
 	for _, chunkDesc := range chunkDescs {
-		wireChunks = append(wireChunks, chunk.NewChunk(userID, fp, metric, chunkDesc.C, chunkDesc.FirstTime, chunkDesc.LastTime))
+		c := chunk.NewChunk(userID, fp, metric, chunkDesc.C, chunkDesc.FirstTime, chunkDesc.LastTime)
+		if err := c.Encode(); err != nil {
```
Comment (gouthamve) [inline pkg/ingester/flush.go]: Is this necessary here, we are calling `Encoded()` in `Put()` now right?

**Comment by gouthamve** [comment]
Ahh, nvm, ignore some of my comments, I realised we passed `chunk.Chunk` and not a reference to it hence the change to `encoded` is not visible. I marked them as resolved. 

Some nits I have not marked resolved still apply. But that is just nitpicking, and this PR looks good otherwise.

File: pkg/chunk/cache/cache_test.go [hunk 5]
```diff
@@ -44,7 +44,9 @@ func fillCache(t *testing.T, cache cache.Cache) ([]string, []chunk.Chunk) {
 			ts.Add(chunkLen),
 		)
 
-		buf, err := c.Encode()
+		err := c.Encode()
```
[the remark this answers is not in the document]
Comment (bboreham) [reply pkg/chunk/cache/cache_test.go]: That would not fit with my stated mission to "Make chunk encoding an explicit operation".

**Pull request merged**